README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# ongoal

<!-- badges: start -->

[![Lifecycle:
experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://lifecycle.r-lib.org/articles/stages.html#experimental)
[![CRAN
status](https://www.r-pkg.org/badges/version/ongoal)](https://CRAN.R-project.org/package=ongoal)
<!-- badges: end -->

ongoal makes 27682 data points available for Pittsburgh Penguins games
where the it was determined whether a shot was on goal (i.e., made it to
the goaltender) or not (i.e., missed or blocked). Data are from seasons
2015-2016, 2016-2017, and 2017-2018.

Several other columns were recorded related to the game, player, and
shot. The data are not perfect; the determination of power play duration
were manually calculated and may not be completely accurate.

## Installation

You can install the development version of ongoal:

``` r
require(pak)
pak::pak("topepo/ongoal")
```

## Example

Here’s a look at the data:

``` r
library(ongoal)
str(on_goal)
#> Classes 'tbl_df', 'tbl' and 'data.frame':    27682 obs. of  24 variables:
#>  $ season             : Factor w/ 3 levels "20152016","20162017",..: 1 1 1 1 1 1 1 1 1 1 ...
#>  $ game_id            : int  2015020010 2015020010 2015020010 2015020010 2015020010 2015020010 2015020010 2015020010 2015020010 2015020010 ...
#>  $ event_idx          : Factor w/ 711 levels "0004","0005",..: 1 3 6 14 16 20 30 38 45 59 ...
#>  $ date_time          : Factor w/ 27832 levels "2015-10-09T00:44:51Z",..: 1 2 3 4 5 6 7 8 9 10 ...
#>  $ event              : Factor w/ 4 levels "Blocked Shot",..: 1 1 4 3 2 3 1 4 1 1 ...
#>  $ on_goal            : Factor w/ 2 levels "yes","no": 2 2 1 2 1 2 2 1 2 2 ...
#>  $ period             : int  1 1 1 1 1 1 1 1 1 1 ...
#>  $ game_seconds       : num  13 36 47 92 99 125 179 220 252 309 ...
#>  $ strength           : Factor w/ 3 levels "even","power_play",..: 1 1 1 1 1 1 1 1 1 1 ...
#>  $ strength_state     : Factor w/ 14 levels "3v3","3v4","3v5",..: 10 10 10 10 10 10 10 10 10 10 ...
#>  $ extra_attacker     : num  0 0 0 0 0 0 0 0 0 0 ...
#>  $ home_skaters       : num  5 5 5 5 5 5 5 5 5 5 ...
#>  $ away_skaters       : num  5 5 5 5 5 5 5 5 5 5 ...
#>  $ goaltender         : Factor w/ 78 levels "aaron_dell","al_montoya",..: 53 11 11 11 53 53 53 53 11 11 ...
#>  $ goal_difference    : num  0 0 0 0 1 1 1 1 -1 -1 ...
#>  $ shooter            : Factor w/ 915 levels "aaron_ekblad",..: 298 875 714 78 577 463 298 699 63 361 ...
#>  $ shooter_team       : Factor w/ 31 levels "ANA","ARI","BOS",..: 10 23 23 23 10 10 10 10 23 23 ...
#>  $ shooter_nationality: Factor w/ 19 levels "AUT","BLR","CAN",..: 15 15 19 19 18 8 15 3 9 3 ...
#>  $ shooter_type       : Factor w/ 5 levels "center","defenseman",..: 1 5 1 5 1 2 1 4 4 4 ...
#>  $ coord_x            : num  -66 -49 64 65 80 42 -55 62 -67 -58 ...
#>  $ coord_y            : num  -11 -21 -31 -21 13 31 -19 15 -9 -16 ...
#>  $ angle              : num  4.06 8.65 51.12 41.19 55.3 ...
#>  $ distance           : num  155.4 139.6 39.8 31.9 15.8 ...
#>  $ behind_goal_line   : num  0 0 0 0 0 0 0 0 0 0 ...
```

There’s also a visualization function (based on [Bill Tran’s
function](https://raw.githubusercontent.com/mrbilltran/the-win-column/master/nhl_rink_plot.R)):

``` r
suppressPackageStartupMessages(library(dplyr))
set.seed(12)
on_goal %>% 
    filter(abs(angle) < 10 & distance < 80) %>% 
    plot_nhl_shots(emphasis = on_goal, alpha = 1 / 2)
```

<img src="man/figures/README-rink-1.png" width="100%" />

## Code of Conduct

Please note that the ongoal project is released with a [Contributor Code
of
Conduct](https://contributor-covenant.org/version/2/0/CODE_OF_CONDUCT.html).
By contributing to this project, you agree to abide by its terms.


NEWS.md
# ongoal 0.0.4

Data refresh after some changes to the hockeyR package. 

# ongoal 0.0.3

New version using the hockeyR package. 

Fixes: 

* Coordinates are changed so that all shots are on the right-hand goal.
* Better (and more official) columns for power plays. 

# ongoal 0.0.2

* Some team data were incorrect and are now fixed. 

* The per-team goals were re-encoded to a new `offense_goal_diff` column.

# ongoal 0.0.1

* Initial version



tests/testthat.R
library(testthat)
library(ongoal)

test_check("ongoal")


tests/testthat/test-rink-plot.R
test_that("rink plots", {
  skip_on_os("windows")
  skip_if_not_installed("dplyr")

  # ----------------------------------------------------------------------------

  save_png <- function(code, file_name = "plot.png",
                       width = 400, height = 400) {
    path <- file.path(tempdir(), file_name)
    png(path, width = width, height = height)
    on.exit(dev.off())
    code

    path
  }

  # ----------------------------------------------------------------------------

  expect_snapshot_file(
    save_png(print(nhl_rink_plot()), "basic-rink-plot.png"),
    name = "basic-rink-plot.png"
  )

  expect_snapshot_file(
    save_png(
      print(nhl_rink_plot("red", "green", "orange")),
      "alt-col-rink-plot.png"
    ),
    name = "alt-col-rink-plot.png"
  )

  # ----------------------------------------------------------------------------

  library(dplyr)
  set.seed(1)
  small_data <- on_goal %>% sample_n(500)

  expect_snapshot_file(
    save_png(
      small_data %>% plot_nhl_shots() %>% print(),
      "basic-data-plot.png"
    ),
    name = "basic-data-plot.png"
  )

  expect_snapshot_file(
    save_png(
      small_data %>% plot_nhl_shots(emphasis = shooter_type) %>% print(),
      "qual-data-plot.png"
    ),
    name = "qual-data-plot.png"
  )

  expect_snapshot_file(
    save_png(
      small_data %>% plot_nhl_shots(emphasis = coord_x) %>% print(),
      "quant-data-plot.png"
    ),
    name = "quant-data-plot.png"
  )

  expect_snapshot_file(
    save_png(
      small_data %>% plot_nhl_shots(cex = 5) %>% print(),
      "point-opt-data-plot.png"
    ),
    name = "point-opt-data-plot.png"
  )

  # ----------------------------------------------------------------------------

  expect_snapshot_error(small_data %>% plot_nhl_shots(emphasis = potato))

})


R/on_goal.R
#' Shots on goal data
#'
#' @description
#' These data are for Pittsburgh Penguins regular season games (seasons
#' 2015-2016, 2016-2017, and 2017-2018). Shots were determined to be on goal
#' (i.e., made it to the goaltender) or not (i.e., missed or blocked).
#'
#' `season_2015` contains data from a single season.
#'
#' @name on_goal
#' @aliases on_goal season_2015
#' @docType data
#' @return \item{on_goal,season_2015}{a tibble}
#' @details
#' The full set of columns are:
#'
#' - `season`: A factor with 3 levels: '20152016', '20162017', '20172018'.
#' - `game_id`, `event_idx`: Index columns for NHL database
#' - `date_time`: Data/time of event
#' - `event`: A factor with 4 levels: 'Blocked Shot', 'Goal', 'Missed Shot',
#'    'Shot'.
#' - `on_goal`: A factor with 2 levels: 'yes', 'no'. An on-goal shot is either
#'    a goal or a shot (from the `event` column).
#' - `period`: (integer) Which of the three regular season periods that the shot
#'    was taken.
#' - `game_seconds`: (double) Cumulative time for the shot.
#' - `strength`: A factor with 3 levels: 'even', 'power_play', 'shorthanded'.
#'    Even means that the teams have equal number of players (which may not be
#'    five). A power play is when the shooting team has more players and
#'    short-handed is when the defensive team (with fewer players).
#' - `strength_state`: A factor with 15 levels with format `#v#`. For example,
#'    "3v5" means the attacking team has three players on the ice and the
#'    defending team has five.
#' - `extra_attacker`: (double) How many more players are on-ice for the
#'    attacking team?
#' - `home_skaters`, `away_skaters`: (double) How many players and on-ice.
#' - `goaltender`: A factor with 78 levels. Name of the current goalie.
#' - `goal_difference`: (double) The difference between the current score of the
#'    attacking team minus the score of the defending team.
#' - `shooter`: A factor with 913 levels. Name of the attacking player.
#' - `shooter_team`: A factor with 31 levels. Attacking team.
#' - `shooter_nationality`: A factor with 19 levels.
#' - `shooter_position`: A factor with 5 levels: 'center', 'defenseman',
#'    'goalie', 'left_wing', or 'right_wing'.
#' - `coord_x`, `coord_y`: (double) Location of the event (shot or blocked
#'    shot). Center ice is a `(0, 0)` so negative x values are on the left and
#'    positive `y` values are on top. The data are configured so that the
#'    attacking team is always shooting at the right-hand goal. Note that
#'    there are not _shot_ locations for blocked shots; these are where the
#'    blocking player was located.
#' - `angle`: (double) The angle of the event (0 to 180 degrees) to the center
#'    of the goal. For blocked shots, this is the angle to the blocking player
#'    (which should approximate shot angle well).
#' - `distance`: (double) The minimum distance from the event to the right-hand
#'    goal (located at `(89, 0)`). We say _minimum_ because, for blocked shots,
#'    it is the distance from the blocking player to the center of the goal.
#' - `behind_goal_line`: (double) For the event, was `x > 89`?
#'
#' The `season_2015` version has fewer columns and different factor levels for
#' the player columns (it is used for teaching).
#' @keywords datasets
#' @examples
#' str(on_goal)
NULL


R/rink-plot.R

#' Plot a hockey rink and add data
#' @param x data frame of data (should include a columns called `coord_x` and
#' `coord_y`).
#' @param emphasis A column in `x` to color (and shape if factor) the data points.
#' If this argument is used, it must be named and cannot use in-line functions
#' (e.g. `emphasis = log(distance)`).
#' @param ... Options to pass to [ggplot2::geom_point()]. If `emphasis` is a
#' string or factor, the `col` and `pch` aesthetics are already set and if
#' `emphasis` is numeric, only `col` is set.
#' @param upward A logical indicating whether the plot should be have center ice
#' at the bottom and the goal near the top. The default is center ice on the left
#' and the goal on the right.
#' @return A ggplot object.
#' @details
#' [nhl_rink_plot()] plots most elements of the rink. Code to plot all can be
#' found in the GitHub link in the References.
#'
#' [plot_nhl_shots()] overlays the rink plot with points in `x`. If `emphasis` is
#' given and is factor or character, the color and shape of the points are set.
#' Otherwise, only the color is set. This function also sets the rink colors to
#' grey.
#'
#' @references \url{https://raw.githubusercontent.com/mrbilltran/the-win-column/master/nhl_rink_plot.R}
#' @author Bill Tran (for `nhl_rink_plot()`)
#' @examples
#' nhl_rink_plot()
#'
#' if (rlang::is_installed("dplyr")) {
#'   library(dplyr)
#'   library(ggplot2)
#'
#'   set.seed(1)
#'   small_data <- on_goal %>% sample_n(500)
#'
#'   small_data %>% filter(distance < 20) %>% plot_nhl_shots()
#'
#'   small_data %>% filter(distance < 50) %>% plot_nhl_shots(alpha = 1/2)
#'
#'   small_data %>%
#'     plot_nhl_shots(emphasis = on_goal, cex = 2, alpha = 1/2) +
#'     scale_color_brewer(palette = "Dark2")
#' }
#'
#'
#' @export
plot_nhl_shots <- function(x, ..., emphasis = NULL, upward = FALSE) {
  if (!all(c("coord_x", "coord_y") %in% names(x))) {
    rlang::abort("data argument 'x' must have columns 'coord_x' and 'coord_y'.")
  }

  p <- nhl_rink_plot(
    NHL_red = grDevices::rgb(0, 0, 0, 0.2),
    NHL_blue = grDevices::rgb(0, 0, 0, 0.2),
    NHL_light_blue = grDevices::rgb(0, 0, 0, 0.2),
    upward = upward
  )

  if (is.null(match.call()$emphasis)) {
    p <- p + ggplot2::geom_point(
      data = x,
      ggplot2::aes(x = coord_x, y = coord_y),
      ...
    )
  } else {
    group_data <- x %>% dplyr::select({{emphasis}}) %>% purrr::pluck(1)
    is_cat <- is.factor(group_data) || is.character(group_data)
    if (is_cat) {
      p <- p + ggplot2::geom_point(
        data = x,
        ggplot2::aes(x = coord_x, y = coord_y, col = {{emphasis}}, pch = {{emphasis}}),
        ...
      )
    } else {
      p <- p + ggplot2::geom_point(
        data = x,
        ggplot2::aes(x = coord_x, y = coord_y, col = {{emphasis}}),
        ...
      )
    }
  }

  p
}

utils::globalVariables(c("coord_x", "coord_y"))

#' @export
#' @rdname plot_nhl_shots
#' @param NHL_red,NHL_blue,NHL_light_blue Suggested colors for various rink elements.
# original from https://raw.githubusercontent.com/mrbilltran/the-win-column/master/nhl_rink_plot.R
nhl_rink_plot <- function (NHL_red = "#FFCCD8", NHL_blue = "#CCE1FF", NHL_light_blue = "#CCF5FF", upward = FALSE) {

  # Plotting an NHL rink completely following the NHL rule book:
  # https://cms.nhl.bamgrid.com/images/assets/binary/308893668/binary-file/file.pdf
  # Line widths, lengths, colours, all followed as closely as possible

  p <-
    ggplot2::ggplot() +
    ggforce::geom_arc(ggplot2::aes(x0 = 0, y0 = 0, start = 0, end = pi, r = 15), col = NHL_blue, alpha = 1/2) +

    # Faceoff circles
    ggforce::geom_circle(ggplot2::aes(x0 = 69, y0 = 22, r = 15), colour = NHL_red, size = 2 / 12) + # Top-Right
    ggforce::geom_circle(ggplot2::aes(x0 = 69, y0 = -22, r = 15), colour = NHL_red, size = 2 / 12) + # Bottom-Right

    # Centre line
    ggplot2::geom_tile(ggplot2::aes(x = 0, y = 0, height = 85), width = 1 / 4, col = NHL_red) + # Centre line

    # Faceoff dots - Plot AFTER centre lines for centre ice circle to show up above
    ggforce::geom_circle(ggplot2::aes(x0 = 0, y0 = 0, r = 6 / 12), colour = NHL_red, fill = NHL_red, size = 0) + # Centre dot with unique red
    ggforce::geom_circle(ggplot2::aes(x0 = 69, y0 = 22, r = 1), colour = NHL_red, fill = NHL_red, size = 0) + # Top-Right
    ggforce::geom_circle(ggplot2::aes(x0 = 69, y0 = -22, r = 1), colour = NHL_red, fill = NHL_red, size = 0) + # Bottom-Right

    ggforce::geom_circle(ggplot2::aes(x0 = 20.5, y0 = 22, r = 1), colour = NHL_red, fill = NHL_red, size = 0) + # Neutral Top-Right
    ggforce::geom_circle(ggplot2::aes(x0 = 20.5, y0 = -22, r = 1), colour = NHL_red, fill = NHL_red, size = 0) + # Neutral Bottom-Right

    # Right goalie crease
    ggplot2::geom_tile(ggplot2::aes(x = 86.75, y = 0, width = 4.5, height = 8), fill = NHL_light_blue) +
    ggforce::geom_arc_bar(ggplot2::aes(x0 = 89, y0 = 0, start = -atan(4.5/4) + 0.01, end = -pi + atan(4.5 / 4) - 0.01, r0 = 4, r = 6), fill = NHL_light_blue, colour = NHL_light_blue, size = 1 / 12) + # manually adjusted arc
    ggplot2::geom_tile(ggplot2::aes(x = 86.75, y = -4, width = 4.5, height = 2 / 12), fill = NHL_red) +
    ggplot2::geom_tile(ggplot2::aes(x = 86.75, y = 4, width = 4.5, height = 2 / 12), fill = NHL_red) +
    ggforce::geom_arc(ggplot2::aes(x0 = 89, y0 = 0, start = -atan(4.5/4) + 0.01, end = -pi + atan(4.5 / 4) - 0.01, r = 6), colour = NHL_red, size = 2 / 12) + # manually adjusted arc
    ggplot2::geom_tile(ggplot2::aes(x = 85, y = 3.75, width = 2 / 12, height = 0.42), fill = NHL_red) +
    ggplot2::geom_tile(ggplot2::aes(x = 85, y = -3.75, width = 2 / 12, height = 0.42), fill = NHL_red) +

    # Goalie nets placed as rectangles
    ggplot2::geom_tile(ggplot2::aes(x = 90.67, y = 0, width = 3.33, height = 6), fill = "#E5E5E3") + # Right

    # Trapezoids
    ggplot2::geom_polygon(ggplot2::aes(x = c(100, 100, 89, 89), y = c(10.92, 11.08, 7.08, 6.92)), fill = NHL_red) + # Right
    ggplot2::geom_polygon(ggplot2::aes(x = c(100, 100, 89, 89), y = c(-10.92, -11.08, -7.08, -6.92)), fill = NHL_red) + # Right

    # Lines
    ggplot2::geom_tile(ggplot2::aes(x = 25.5, y = 0, width = 1, height = 85),  fill = NHL_blue) + # Right Blue line
    ggplot2::geom_tile(ggplot2::aes(x = 89, y = 0, width = 2 / 12, height = 73.50), fill = NHL_red) + # Right goal line

    # Borders as line segments - plotted last to cover up line ends, etc.
    ggplot2::geom_line(ggplot2::aes(x = c(0, 72), y = c(42.5, 42.5))) + # Top
    ggplot2::geom_line(ggplot2::aes(x = c(0, 72), y = c(-42.5, -42.5))) + # Bottom
    ggplot2::geom_line(ggplot2::aes(x = c(100, 100), y = c(-14.5, 14.5))) + # Right
    ggforce::geom_arc(ggplot2::aes(x0 = 72, y0 = 14.5, start = pi / 2, end = 0, r = 28)) + # Top-Right
    ggforce::geom_arc(ggplot2::aes(x0 = 72, y0 = -14.5, start = pi, end =  pi / 2, r = 28)) + # Bottom-Right

    ggplot2::theme_void() +
    ggplot2::theme(legend.position = "top") +
    ggplot2::theme(
      axis.title.x = ggplot2::element_blank(),
      axis.text.x  = ggplot2::element_blank(),
      axis.ticks.x = ggplot2::element_blank(),
      axis.title.y = ggplot2::element_blank(),
      axis.text.y  = ggplot2::element_blank(),
      axis.ticks.y = ggplot2::element_blank(),
      panel.grid.major = ggplot2::element_blank(),
      panel.grid.minor = ggplot2::element_blank(),
      panel.border = ggplot2::element_blank()
    )

  if (upward) {
    p <- p + ggplot2::coord_flip()
  }
  p
}


inst/import_data.R
# pak::pak(c("danmorse314/hockeyR"), ask = FALSE)

library(hockeyR)
library(nhlapi) # for player position
library(tidyverse)
library(tidylog, warn.conflicts = FALSE)
library(sessioninfo)
library(usethis)

# ------------------------------------------------------------------------------

play_by_play <- bind_rows(load_pbp('2015-16'), load_pbp('2016-17'), load_pbp('2017-18'))

# ------------------------------------------------------------------------------

clean_values <- function(x) {
  x <- gsub("([[:punct:]])|([[:space:]])", "_", tolower(x))
  x <- gsub("__", "_", x) # for names with dot them space
  x
}

# ------------------------------------------------------------------------------
# Basic filtering of events and teams along with some simple feature creation or
# engineering

shots <-
  play_by_play  %>%
  # Filter for shot-related events
  filter(event %in% c("Shot", "Blocked Shot", "Missed Shot", "Goal")) %>%
  # Remove 0v0 and 0v5 ¯\_(ツ)_/¯
  filter(!grepl("^0v", strength_state)) %>%
  # Remove pre-season games and overtime shots
  filter(period_type == "REGULAR" & season_type == "R") %>% 
  select(
    event, event_team_type, event_team_abbr, period, game_seconds, 
    home_score, away_score,
    # The event_player_* cols will be used to get info on the shooter
    starts_with("event_player_"), -ends_with("_link"),  -ends_with("_season_total"),
    strength, strength_state, extra_attacker, empty_net,
    x_fixed, y_fixed, home_skaters, away_skaters, ends_with("_goalie"),
    date_time, event_team_abbr, season, event_idx, game_id
  ) %>% 
  relocate(season, game_id, event_idx, date_time)

# Filter data down via including games played by PIT
pit_playing <-
  shots %>%
  group_by(season, game_id) %>%
  summarize(has_pit = any(event_team_abbr == "PIT"), .groups = "drop") %>% 
  filter(has_pit) %>%
  select(-has_pit)

team_info <-
  shots %>%
  # Next line to tie shooter to team; see notes on blocked shots below
  filter(event == "Shot") %>% 
  group_by(season, game_id) %>%
  dplyr::distinct(event_team_type, event_team_abbr) %>%
  pivot_wider(
    id_cols = c(season, game_id),
    names_from = event_team_type,
    values_from = event_team_abbr
  ) %>% 
  ungroup()

pit_playing <- inner_join(pit_playing, team_info, by = c("season", "game_id"))

pit_shots <- 
  left_join(pit_playing, shots, by = c("season", "game_id")) %>% 
  mutate(
    # 'event_team_abbr' is the defensive team for a blocked shot. 
    # If that is the case, switch the team (as well as 'event_team_type')
    other_team = ifelse(event_team_abbr == away, home, away),
    event_team_abbr = ifelse(event == "Blocked Shot", other_team, event_team_abbr),
    other_type = ifelse(event_team_type == "away", "home", "away"),
    event_team_type = ifelse(event == "Blocked Shot", other_type, event_team_type),
    
    # Convert logical to binary indicator
    extra_attacker = ifelse(extra_attacker, 1, 0),
    
    # Assign no goaltender when an empty net
    empty_net = isTRUE(empty_net),
    away_goalie = ifelse(is.na(away_goalie) & empty_net, "none", away_goalie),
    home_goalie = ifelse(is.na(home_goalie) & empty_net, "none", home_goalie),
    
    # Determine goaltender from opposite team as event. There is a fair amount
    # of missing data here for some games. See the code below where we use the
    # 'event_player_*' columns to try to get the data when these two columns
    # are missing. 
    goaltender = ifelse(event_team_type == "home", away_goalie, home_goalie),

    # Make a column for goal differential between offense and defense team
    goal_difference = ifelse(event_team_type == "away", 
                             away_score - home_score, 
                             home_score - away_score),
    
    # Binary outcome
    on_goal = ifelse(event %in% c("Shot", "Goal"), "yes", "no"),
    on_goal = factor(on_goal, levels = c("yes", "no"))
  ) %>% 
  rename(shooter_team = event_team_abbr, team_away = away, team_home = home) %>% 
  select(-other_team, -other_type, -away_score, -home_score) %>% 
  relocate(event, on_goal, .after = date_time)

# ------------------------------------------------------------------------------
# Get shooter name and position. Also get goaltender info to replace missing 
# values of 'away_goalie' or 'home_goalie'

players <-
  pit_shots %>%
  select(event_idx, game_id, season, starts_with("event_player_")) %>%
  # have to convert integer player ID to character :-O
  mutate(across(c(starts_with("event_player_")), ~ as.character(.x))) %>%
  pivot_longer(
    c(starts_with("event_player_")),
    names_to = c("col"),
    values_to = "data"
  ) %>%
  mutate(
    event_index = gsub("([[:alpha:]])|([[:punct:]])", "", col),
    col = gsub("event_player_(.*)_", "", col)
  ) %>%
  pivot_wider(
    id_cols = c(season, game_id, event_idx, event_index),
    names_from = col,
    values_from = data
  ) %>%
  filter(!is.na(name))

shooter <- 
  players %>%
  filter(type %in%  c("Shooter", "Scorer")) %>%
  mutate(
    name = clean_values(name),
    id = as.integer(id)
  ) %>%
  select(season, game_id, event_idx, shooter = name, shooter_id = id)

goalie <- 
  players %>%
  filter(type == "Goalie") %>%
  select(season, game_id, event_idx, event_goalie = name)


# use nhlapi to get shooter position

shooter_ids <-
  shooter %>%
  dplyr::distinct(shooter_id) %>%
  pluck("shooter_id")

shooter_type <-
  nhl_players(playerIds = shooter_ids) %>%
  mutate(shooter_type = gsub(" ", "_", tolower(primaryPosition.name))) %>%
  select(shooter_id = id, shooter_nationality = nationality, shooter_type) %>%
  filter(!is.na(shooter_type))

shooter_with_position <-
  inner_join(shooter, shooter_type, by = "shooter_id") %>%
  select(-shooter_id)

with_shooter <-
  pit_shots %>%
  left_join(shooter_with_position, by = c("season", "game_id", "event_idx")) %>%
  select(-starts_with("event_player_")) %>% 
  # There are a handful of rows without shooter info
  filter(!is.na(shooter)) %>% 
  relocate(shooter_team, .after = shooter)

with_goalie <-
  with_shooter %>% 
  left_join(goalie, by = c("season", "game_id", "event_idx")) %>% 
  mutate(
    goaltender = ifelse(is.na(goaltender) & !is.na(event_goalie), event_goalie, goaltender),
    goaltender = clean_values(goaltender)
  )

# ------------------------------------------------------------------------------
# Make all shots on right-hand goal. From hockeyR docs: "x_fixed: Numeric 
# transformed x-coordinate of event in feet, where the home team always shoots 
# to the right, away team to the left." 

left_to_right <- 
  with_goalie %>% 
  mutate(
    coord_x = ifelse(event_team_type == "away", -x_fixed, x_fixed),
    coord_y = ifelse(event_team_type == "away", -y_fixed, y_fixed)
  )

# ------------------------------------------------------------------------------

on_goal <- 
  left_to_right %>% 
  select(-x_fixed, -y_fixed, -event_team_type) %>% 
  mutate(
    strength = clean_values(strength),
    shooter = clean_values(shooter),
    across(where(is.character), as.factor),
    angle = abs( atan2(abs(coord_y), (89 - coord_x) ) * (180 / pi)),
    distance = sqrt( (89 - coord_x) ^ 2 + coord_y ^ 2 ), 
    behind_goal_line = ifelse(coord_x >= 89, 1, 0)
  ) %>% 
  select(-team_home, -team_away, -empty_net, -home_goalie, -away_goalie, 
         -event_goalie)


on_goal <- on_goal[complete.cases(on_goal), ]

# ------------------------------------------------------------------------------
# Smaller version used for teaching

retain <-
  c("on_goal", "period", "game_seconds", "strength", "home_skaters",
    "away_skaters", "goaltender", "goal_difference", "shooter", "shooter_type",
    "coord_x", "coord_y", "extra_attacker")

season_2015 <-
  on_goal %>%
  filter(season == "20152016") %>%
  select(all_of(retain)) %>%
  dplyr::mutate(
    shooter = factor(as.character(shooter)),
    goaltender = factor(as.character(goaltender))
  )

use_data(
  on_goal,
  version = 3,
  overwrite = TRUE,
  compress = TRUE
)

use_data(
  season_2015,
  version = 3,
  overwrite = TRUE,
  compress = TRUE
)

# ------------------------------------------------------------------------------

if (!interactive()) {
  q("no")
}
